File: src/main/java/dev/jlo/gamemodes/paper/PaperGameplayListener.java
package dev.jlo.gamemodes.paper;

import dev.jlo.gamemodes.domain.common.MatchPhase;
import java.util.UUID;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.EntityDamageByEntityEvent;
import org.bukkit.event.entity.PlayerDeathEvent;
import org.bukkit.event.player.PlayerJoinEvent;
import org.bukkit.event.player.PlayerMoveEvent;
import org.bukkit.event.player.PlayerQuitEvent;
import org.bukkit.event.player.PlayerRespawnEvent;

public final class PaperGameplayListener implements Listener {
    private final PaperMatchService service;
    private final PaperUi ui;

    public PaperGameplayListener(PaperMatchService service, PaperUi ui) {
        this.service = service;
        this.ui = ui;
    }

    @EventHandler
    public void onJoin(PlayerJoinEvent event) {
        service.restore(event.getPlayer());
    }

    @EventHandler
    public void onQuit(PlayerQuitEvent event) {
        service.disconnect(event.getPlayer().getUniqueId());
        ui.clear(event.getPlayer());
    }

    @EventHandler
    public void onRespawn(PlayerRespawnEvent event) {
        var location = service.respawn(event.getPlayer());
        if (location != null) {
            event.setRespawnLocation(location);
        }
    }

    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    public void onMove(PlayerMoveEvent event) {
        if (!event.hasChangedPosition()) {
            return;
        }
        MatchSession session = null;
        UUID playerId = event.getPlayer().getUniqueId();
        for (MatchSession candidate : service.sessions()) {
            if (candidate.teamOf(playerId) != null) {
                session = candidate;
                break;
            }
        }
        if (session == null) {
            return;
        }
        ui.update(event.getPlayer(), session);
        session.reconcileObjective(playerId, event.getPlayer().getLocation());
    }

    @EventHandler
    public void onDeath(PlayerDeathEvent event) {
        UUID playerId = event.getPlayer().getUniqueId();
        if (!service.isParticipant(playerId)) {
            return;
        }
        event.setKeepInventory(false);
        Player killer = event.getPlayer().getKiller();
        service.playerDied(playerId, killer == null ? null : killer.getUniqueId());
    }

    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    public void onDamage(EntityDamageByEntityEvent event) {
        if (!(event.getEntity() instanceof Player victim)) {
            return;
        }
        if (!(event.getDamager() instanceof Player killer)) {
            return;
        }
        MatchSession session = null;
        UUID victimId = victim.getUniqueId();
        UUID killerId = killer.getUniqueId();
        for (MatchSession candidate : service.sessions()) {
            if (candidate.getPhase() == MatchPhase.ACTIVE
                    && candidate.teamOf(victimId) != null
                    && candidate.teamOf(killerId) != null) {
                session = candidate;
                break;
            }
        }
        if (session == null) {
            return;
        }
        if (session.teamOf(killerId) == session.teamOf(victimId)) {
            event.setCancelled(true);
        } else {
            service.recordCombat(killerId, victimId);
        }
    }
}

File: src/main/kotlin/dev/jlo/gamemodes/paper/PaperGameplayListener.kt
package dev.jlo.gamemodes.paper

import dev.jlo.gamemodes.domain.common.MatchPhase
import dev.jlo.gamemodes.domain.opr.OutpostId
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.entity.EntityDamageByEntityEvent
import org.bukkit.event.entity.PlayerDeathEvent
import org.bukkit.event.player.PlayerJoinEvent
import org.bukkit.event.player.PlayerMoveEvent
import org.bukkit.event.player.PlayerQuitEvent
import org.bukkit.event.player.PlayerRespawnEvent

class PaperGameplayListener(
    private val service: PaperMatchService,
    private val ui: PaperUi
) : Listener {
    @EventHandler
    fun onJoin(event: PlayerJoinEvent) { service.restore(event.player) }

    @EventHandler
    fun onQuit(event: PlayerQuitEvent) {
        service.disconnect(event.player.uniqueId)
        ui.clear(event.player)
    }

    @EventHandler
    fun onRespawn(event: PlayerRespawnEvent) {
        service.respawn(event.player)?.let { event.respawnLocation = it }
    }

    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    fun onMove(event: PlayerMoveEvent) {
        if (!event.hasChangedPosition()) return
        val session = service.sessions().firstOrNull { it.teamOf(event.player.uniqueId) != null } ?: return
        ui.update(event.player, session)
        session.reconcileObjective(event.player.uniqueId, event.player.location)
    }

    @EventHandler
    fun onDeath(event: PlayerDeathEvent) {
        if (!service.isParticipant(event.player.uniqueId)) return
        event.keepInventory = false
        service.playerDied(event.player.uniqueId, event.player.killer?.uniqueId)
    }

    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    fun onDamage(event: EntityDamageByEntityEvent) {
        val victim = event.entity as? Player ?: return
        val killer = event.damager as? Player ?: return
        val session = service.sessions().firstOrNull {
            it.phase == MatchPhase.ACTIVE &&
                it.teamOf(victim.uniqueId) != null &&
                it.teamOf(killer.uniqueId) != null
        } ?: return
        if (session.teamOf(killer.uniqueId) == session.teamOf(victim.uniqueId)) {
            event.isCancelled = true
        } else {
            service.recordCombat(killer.uniqueId, victim.uniqueId)
        }
    }
}
